app/common/collections/runner.py

```diff
@@ -0,0 +1,179 @@
+        if not self.question:
+            raise RuntimeError("Question context not set")
+        return self._question_form
+
+    @property
+    def check_your_answers_form(self) -> "CheckYourAnswersForm":
+        if not self.form:
+            raise RuntimeError("Form context not set")
+        return self._check_your_answers_form
+
+    @property
+    def tasklist_form(self) -> "GenericSubmitForm":
+        return self._tasklist_form
+
+    def save_question_answer(self) -> None:
+        if not self.question:
+            raise RuntimeError("Question context not set")
+
+        self.submission.submit_answer_for_question(self.question.id, self.question_form)
+
+    def save_is_form_completed(self, user: "User") -> None:
+        if not self.form:
+            raise RuntimeError("Form context not set")
+
+        try:
+            self.submission.toggle_form_completed(
+                form=self.form,
+                user=user,
+                is_complete=self.check_your_answers_form.section_completed.data == "yes",
+            )
+        except ValueError as e:
+            self.check_your_answers_form.section_completed.errors.append(  # type:ignore[attr-defined]
+                "You must complete all questions before marking this section as complete"
+            )
+            raise ValueError("Failed to save form completed status") from e
+
+    def submit(self, user: "User") -> None:
+        try:
+            self.submission.submit(user)
+            notification_service.send_collection_submission(self.submission.submission)
+        except ValueError as e:
+            self._tasklist_form.submit.errors.append("You must complete all forms before submitting the collection")  # type:ignore[attr-defined]
+            raise ValueError("Failed to submit") from e
+
+    def to_url(
+        self,
+        state: FormRunnerState,
+        *,
+        question: Optional["Question"] = None,
+        form: Optional["Form"] = None,
+        source: Optional[FormRunnerState] = None,
+    ) -> str:
+        return self.url_map[state](self, question or self.question, form or self.form, source)
+
+    @property
+    def next_url(self) -> str:
+        if self.question and self._valid is not None:
+            if self._valid:
+                if self.source == FormRunnerState.CHECK_YOUR_ANSWERS:
+                    # todo: even if we came from check your answers, move to the next question
+                    #       if it hasn't been answered
+                    return self.to_url(FormRunnerState.CHECK_YOUR_ANSWERS)
+
+                next_question = self.submission.get_next_question(self.question.id)
+                if next_question:
+                    return self.to_url(FormRunnerState.QUESTION, question=next_question)
+
+                return self.to_url(FormRunnerState.CHECK_YOUR_ANSWERS)
+            else:
+                return self.to_url(FormRunnerState.CHECK_YOUR_ANSWERS)
+
+        return self.to_url(FormRunnerState.TASKLIST)
```

Comment:
I think the routing logic here is a bit complex - maybe comments on each bit specifying the scenario would be useful?